File: macos/Gutenkern/Sources/GutenkernCore/KerningHitTesting.swift
import Foundation

public struct PairHitFrame: Equatable, Sendable {
    public var key: String
    public var x: Double
    public var y: Double
    public var width: Double
    public var height: Double

    public init(key: String, x: Double, y: Double, width: Double, height: Double) {
        self.key = key
        self.x = x
        self.y = y
        self.width = width
        self.height = height
    }
}

public enum PairHitTesting {
    public static func nearestKey(x: Double, y: Double, frames: [PairHitFrame]) -> String? {
        var bestKey: String?
        var bestDistance = Double.infinity
        for frame in frames {
            let distance = distanceToRect(x: x, y: y, frame: frame)
            if distance < bestDistance {
                bestDistance = distance
                bestKey = frame.key
            }
        }
        return bestKey
    }

    private static func distanceToRect(x: Double, y: Double, frame: PairHitFrame) -> Double {
        let minX = frame.x
        let minY = frame.y
        let maxX = frame.x + frame.width
        let maxY = frame.y + frame.height
        let dx: Double
        if x < minX {
            dx = minX - x
        } else if x > maxX {
            dx = x - maxX
        } else {
            dx = 0
        }
        let dy: Double
        if y < minY {
            dy = minY - y
        } else if y > maxY {
            dy = y - maxY
        } else {
            dy = 0
        }
        return (dx * dx + dy * dy).squareRoot()
    }
}

File: windows/Gutenkern.Core/KerningHitTesting.cs
namespace Gutenkern;

public readonly struct PairHitFrame
{
    public string Key { get; }
    public double X { get; }
    public double Y { get; }
    public double Width { get; }
    public double Height { get; }

    public PairHitFrame(string key, double x, double y, double width, double height)
    {
        Key = key;
        X = x;
        Y = y;
        Width = width;
        Height = height;
    }
}

public static class PairHitTesting
{
    public static string? NearestKey(double x, double y, IReadOnlyList<PairHitFrame> frames)
    {
        string? bestKey = null;
        var bestDistance = double.PositiveInfinity;
        foreach (var frame in frames)
        {
            var distance = DistanceToRect(x, y, frame);
            if (distance < bestDistance)
            {
                bestDistance = distance;
                bestKey = frame.Key;
            }
        }

        return bestKey;
    }

    private static double DistanceToRect(double x, double y, PairHitFrame frame)
    {
        var minX = frame.X;
        var minY = frame.Y;
        var maxX = frame.X + frame.Width;
        var maxY = frame.Y + frame.Height;
        var dx = 0.0;
        if (x < minX)
        {
            dx = minX - x;
        }
        else if (x > maxX)
        {
            dx = x - maxX;
        }

        var dy = 0.0;
        if (y < minY)
        {
            dy = minY - y;
        }
        else if (y > maxY)
        {
            dy = y - maxY;
        }

        return Math.Sqrt(dx * dx + dy * dy);
    }
}
